### PR TITLE
add metadata tags [DB SCHEMA CHANGE]

### PR DESCRIPTION
**Pull request by diegocepedaw**
add the ability to tag applications, plans, templates, incidents, and targets with config defined metadata

### **WARNING: this commit adds new table to the mysql schema that existing routes will use in their queries. You must create the new metadata_tag tables before deploying this version!!!**

TODO:
- add full filtering capability to targets endpoint
- update the sync script and retention to also delete tags for resources they clean up

### CODE REVIEW

File: src/iris/api.py [hunk 1]
```diff
@@ -6662,6 +6846,189 @@ def on_get(self, req, resp):
...
+        raise HTTPBadRequest('Invalid request: name must be string')
+    if name not in allowed_tags:
+        raise HTTPBadRequest('Invalid request: name is not one of the allowlisted tag names: %s' % str(allowed_tags))
+    if not (isinstance(values, list) and all(isinstance(value, str) for value in values)):
+        raise HTTPBadRequest('Invalid request: values must a list of strings')
+
+    return (name, values, id)
+
+
+class IncidentTag():
```
[the remark this answers is not in the document]
Comment (diegocepedaw) [reply src/iris/api.py:6872]: I did try to share as much as possible and considered this approach as well but I wasn't sure if what I had or this approach would be cleaner. We do have to pass 4 different strings (allowed_tags key, resource table name, tag table name, resource id field name) to have the queries work for the correct type of resource but that is definitely viable as well if we want to go that direction. I'll try out that change

File: src/iris/api.py [hunk 2]
```diff
@@ -1244,15 +1323,36 @@ def on_get(self, req, resp):
...
 
+        tag_subquery = gen_tag_where_subquery(connection, 'plan_id', 'plan_metadata_tag', req.params)
+        if tag_subquery != "":
+            where.append(tag_subquery)
+
         if where:
             query = query + ' WHERE ' + ' AND '.join(where)
 
+        # since we are joining with a table that has many to one relationship apply a group by to remove duplicates
+        query += ' GROUP BY `plan`.`id`'
```
[the remark this answers is not in the document]
Comment (diegocepedaw) [reply src/iris/api.py]: good idea, made the change

File: src/iris/api.py (hunk 1 above)
[the remark this answers is not in the document]
Comment (jrgp) [reply src/iris/api.py]: > We do have to pass 4 different strings (allowed_tags key, resource table name, tag table name, resource id field name) to have the queries work for the correct type of resource which felt a bit weirdly un-idiomatic

Ah yeah, my idea was if we could use string concatenation to come up with the correct names of tables/allowed tags key/etc. 

If that is too hacky though we can go with the original approach

[the remark this answers is not in the document]
Comment (diegocepedaw) [reply src/iris/api.py]: changed

File: src/iris/api.py [hunk 3]
```diff
@@ -729,6 +740,57 @@ def is_valid_tracking_settings(t, k, tpl):
...
+
+    if len(conditions) == 0:
+        return ""
+
+    # Combine all conditions using AND to ensure they are all satisfied
+    combined_conditions = " AND ".join(conditions)
+    # Combine all joins
+    combined_joins = " ".join(joins)
+    # SQL query to retrieve distinct resource ids based on the conditions and self-joins
+    sql_query = f"EXISTS (SELECT t0.{id_field} FROM {tag_table} AS t0 {combined_joins} WHERE {combined_conditions} AND t0.{id_field} = {resource_id})"
```
Comment (jrgp) [inline src/iris/api.py:790]: With a `exists(select` query, we don't need to include an actual field name in the SELECT clause as it gets ignored 

Eg we can just do `EXISTS (SELECT 1 FROM `

Comment (diegocepedaw) [reply src/iris/api.py]: changed

**Review by jrgp — APPROVED** [review]

**Pull request merged**